miniapps/meshing/mesh-optimizer.hpp

```diff
@@ -279,3 +386,959 @@ void DiffuseField(ParGridFunction &field, int smooth_steps)
    delete Lap;
 }
 #endif
+
+
+/// This class is used to update GridFunction and FESpaces after mesh updates.
+/// It also provides functionality to reset essential boundary conditions, and
+/// rebalance ParMesh.
+class TMOPAMR
+{
+protected:
+   Mesh *mesh;
+   NonlinearForm *nlf;
+   Array<GridFunction *> gridfuncarr;
+   Array<FiniteElementSpace *> fespacearr;
+#ifdef MFEM_USE_MPI
+   ParMesh *pmesh;
+   ParNonlinearForm *pnlf;
+   Array<ParGridFunction *> pgridfuncarr;
+   Array<ParFiniteElementSpace *> pfespacearr;
+#endif
+
+   bool move_bnd;
+
+public:
+   TMOPAMR(Mesh &mesh_, NonlinearForm &nlf_, bool move_bnd_) :
+      mesh(&mesh_), nlf(&nlf_), gridfuncarr(), fespacearr(), move_bnd(move_bnd_) { }
+#ifdef MFEM_USE_MPI
+   TMOPAMR(ParMesh &pmesh_, ParNonlinearForm &pnlf_, bool move_bnd_) :
+      mesh(&pmesh_), nlf(&pnlf_), gridfuncarr(), fespacearr(),
+      pmesh(&pmesh_), pnlf(&pnlf_), pgridfuncarr(), pfespacearr(),
+      move_bnd(move_bnd_) { }
+#endif
+
+   void AddGridFunctionForUpdate(GridFunction *gf_) { gridfuncarr.Append(gf_); }
+#ifdef MFEM_USE_MPI
+   void AddGridFunctionForUpdate(ParGridFunction *pgf_) { pgridfuncarr.Append(pgf_); }
+#endif
+   void AddFESpaceForUpdate(FiniteElementSpace *fes_) { fespacearr.Append(fes_); }
+#ifdef MFEM_USE_MPI
+   void AddFESpaceForUpdate(ParFiniteElementSpace *pfes_) { pfespacearr.Append(pfes_); }
+#endif
+
+   void Update();
+#ifdef MFEM_USE_MPI
+   void ParUpdate();
+#endif
+
+#ifdef MFEM_USE_MPI
+   // Rebalance ParMesh such that all the children elements are moved to the same
+   // MPI rank where the parent will be if the mesh were to be derefined.
+   void RebalanceParNCMesh();
+#endif
+};
+
+#ifdef MFEM_USE_MPI
+void TMOPAMR::RebalanceParNCMesh()
+{
+   ParNCMesh *pncmesh = pmesh->pncmesh;
+   if (pncmesh)
+   {
+      const Table &dreftable = pncmesh->GetDerefinementTable();
+      Array<int> drefs, new_ranks;
+      for (int i = 0; i < dreftable.Size(); i++)
+      {
+         drefs.Append(i);
+      }
+      pncmesh->GetFineToCoarsePartitioning(drefs, new_ranks);
+      new_ranks.SetSize(pmesh->GetNE());
```

Comment:
Doesn't `new_ranks` come out of the `GetFineToCoarsePartitioning` call (above) with the correct size?